### DISCUSSION TITLE
What does safeBackup do - or what triggers it?

### DISCUSSION THREAD

**Discussion by gunner007cd** [abraunegg/onedrive#3027] [Q&A] [answered]
Pretty much the title. I have OneDrive personal running in a docker and it's been running fine through the years before updates. However, recently, there's one file that its been creating FileName-safeBackup-001.dat and has been creating them every time the file has changed. Now there's about 100 safeBackup files in there. Even if I delete them all and delete the original file and upload the new one, the docker will synchronize and then about 5 minutes later, will create another safeBackup.dat file.

Is there something wrong with the configuration or a way to suppress the creation of files?

**Comment by abraunegg** [comment, marked as answer]
@gunner007cd 
> What does safeBackup do

It backs up your file as it has changed outside the recorded online|local changes, outside of the normal sync process, normal use of the client.

> or what triggers it?

Firstly, ask yourself how are you using the client:
1. Are you *always* using `--resync` which is the main contributor here to performing a safe backup of your data. If this is the case - stop using `--resync` unless the client actually tells you to use it.
2. Enable verbose logging with `--verbose` to gain insight as to when these files are being created

Using the verbose logs, please read the documentation: https://github.com/abraunegg/onedrive/blob/master/docs/client-architecture.md to help you understand why they are being created.

Look at all the sequence diagrams where data loss prevention is being flagged.

> Is there something wrong with the configuration or a way to suppress the creation of files?

Look at your usage first, it is odd that the client is constantly doing this for one particular file.

If you want to bypass data preservation, this is possible, but this should be a last resort item:

https://github.com/abraunegg/onedrive/blob/master/docs/application-config-options.md#bypass_data_preservation

Please read *all* the documentation.

**Reply by gunner007cd** [reply]
I don't offhand remember using the --resync function when creating it, but perhaps at one point I may have used it to do a sync when I had lost internet connection. I will grab the verbose log, it always seemed to be anywhere from 5 minutes to several hours. 

The only lines in my config file were

sync_dir = "/onedrive/data"
monitor_interval = "3600"
monitor_fullscan_frequency = "24"

Which basically I was telling it to sync once per hour and doing a full scan every 24 hours.

Since there's not a ton of files, I deleted the docker and just recreated it with the config file above - no resync usage. I'll check to see if any files have SafeBackup.dat anymore.

**Reply by abraunegg** [reply]
So your configuration is at odds with your experience and when the files are being created.

If they are being created ever 5 mins - then you potentially have 2 onedrive processes maybe hitting the same data and causing a conflict .. unless your local filesystem is triggering an inotify event for file changed, and the backup occuring that way - however the issue here is safe backup is used when pulling down content from online, and then the conflict occurs.

I would be looking at seeing if you have any duplicate processes or containers running with the default configuration somehow

**Reply by gunner007cd** [reply]
Hmm. I have the Onedrive Windows client installed on my desktop only. However, now that I think about it - it's related to a quicken file that's in OneDrive. Is it possible I'm just running into a random scenario where it's log file just so happens to be open during the monitor interval?

**Reply by abraunegg** [reply]
Enabling verbose logging and/or sending the logging to a separate file for post analysis is the best advice I can give you. You can then corelate the file name with the relevant path and or your activity at that point identified by the timestamp.

**Reply by gunner007cd** [reply]
Yup have it enabled now - will follow up if I see it crop up again - thank you for your assistance!

**Reply by gunner007cd** [reply]
It was what I expected, it's happening if that specific program just happens to be open at the time they monitor interval goes off. I think I understand it now. Thank you.

**Reply by mireianievas** [reply]
I also got a directory completely messed up with these safeBackup files. Can it be disabled somehow? or how can I remove them? beause the owner of the repo cannot, and I can just one-by-one remove them (they are hundreds of files ... ) . I have not used resync lately. But I was getting sync errors before (error code 141) and I just updated curl to try to solve those.

**Reply by abraunegg** [reply]
@mireianievas 
> Can it be disabled somehow? 

Please read this discussion and/or read the documentation and understand the consequence of disabling this capability

> or how can I remove them?

You can manually search for files that contain `safeBackup` and delete them - but remember, these are files that has changed content - if you simply delete them without review - you may experience data loss .. and this is *why* they exist.

> But I was getting sync errors before (error code 141) and I just updated curl to try to solve those.

If you are exit code 141 - you are most likely hitting this: https://github.com/abraunegg/onedrive/blob/master/docs/known-issues.md#application-stops-running-without-any-visible-reason

This is due to you using a distribution with a broken 'curl' version - please read: 
* https://github.com/abraunegg/onedrive/blob/master/docs/usage.md#compatibility-with-curl
* https://github.com/abraunegg/onedrive/discussions/2997

If you are using Ubuntu .. your choice, but there are better distributions out there. 

Please also ensure you install the client the *correct* way - especially on Ubuntu (do not use the Universe package as it is broken)

**Reply by graealex** [reply]
> It backs up your file as it has changed outside the recorded online|local changes, outside of the normal sync process, normal use of the client.

You are mistaken. Those files appear when your OneDrive sync application shits itself. That's all. All it would take is to compare the supposedly "new" file with the old file to figure out that literally nothing has changed, and no "xxxx-safeBackup-0001" had to be created. Especially since those files are completely isolated, i.e. no original file is preserved on the disk.

It's a bug. Stop being so vain.

**Reply by graealex** [reply]
> It backs up your file as it has changed outside the recorded online|local changes, outside of the normal sync process, normal use of the client.

<img width="1157" height="396" alt="image" src="https://github.com/user-attachments/assets/056bbc27-29ba-4687-8c24-3d30c3d8cc57" />

How likely do you think it is that a ripped FLAC file from a CD has spontaneously changed?

**Reply by abraunegg** [reply]
@graealex 
If you think there is a bug - get the evidence and raise a bug report.

What 99.99% of users do not realiae is that Microsoft willfully modifies your file online by adding metadata and changes the timestamp that fundamentally changes the file itself.

**Reply by graealex** [reply]
> What 99.99% of users do not realiae is that Microsoft willfully modifies your file online by adding metadata and changes the timestamp that fundamentally changes the file itself.

How does changed metadata (attributes, modified, created date) constitute the need for a "xxxx-safeBackup-xxx" to be created? Again, my question was, how much bytes do you estimate have been changed in those particular FLAC files? The content stream itself.

**Reply by abraunegg** [reply]
@graealex 

The metadata *changes* the hash of the file.

Please read: https://github.com/abraunegg/onedrive/blob/master/docs/usage.md#why-does-the-client-create-safebackup-files

**Reply by graealex** [reply]
See, that's a major flaw with the software. At least if you then keep pretending it's two different files, when only metadata has changed, and subsequently you create a "xxxx-safeBackup-xxx" file. And nothing of that explains why there's ONLY the "xxxx-safeBackup-xxx" in the folder, instead of two copies.

**Reply by abraunegg** [reply]
@graealex 

> See, that's a major flaw with the software. At least if you then keep pretending it's two different files, when only metadata has changed, and subsequently you create a "xxxx-safeBackup-xxx" file. And nothing of that explains why there's ONLY the "xxxx-safeBackup-xxx" in the folder, instead of two copies.


Again .. *if* you believe that there is a bug, follow the support process and raise a bug: https://github.com/abraunegg/onedrive#reporting-an-issue-or-bug

**Reply by graealex** [reply]
Nah, I'm not going through the motions, sorry. The software sucks, my NAS is filled with safeBackup files. Maybe just compare contents and don't pretend a modified date is part of it. You've explained that metadata goes into the hash, which is simply wrong. Hash the content, and compare metadata separately.

**Comment by peterborto** [comment]
Hi  @abraunegg  i've too the problem with safe backup folders, every 2/3 days it create a new backup.

In my config i do not have  resync enabled.

...
Config option 'resync'                       = false
Config option 'resync_auth'                  = false
...

May i have to look in another location ?

**Reply by abraunegg** [reply]
@peterborto 
My suggestion is for you to run the client in verbose mode so you can then understand why the files are being created.

Please also ensure:
* you have a current version of curl / libcurl installed
* you have v2.5.5 or greater as your application version

**Reply by peterborto** [reply]
Here info:

i suspect problem is with one drive gui (https://github.com/bpozdena/OneDriveGUI this project)

Application version                          = onedrive v2.5.5
Compiled with                                = LDC 2110
Curl version                                 = libcurl/8.11.1 OpenSSL/3.2.4 zlib/1.3.1.zlib-ng brotli/1.1.0 libidn2/2.3.8 libpsl/0.21.5 libssh/0.11.1/openssl/zlib nghttp2/1.64.0 OpenLDAP/2.6.9

I have enabled debug log, i hope i will find something more info to put there.

**Reply by peterborto** [reply]
found issue: 
```
2025-05-12 08:58:02,616 [OneDriveGUI.py:2022][fn=read_stdout][INFO] - [my user] _**This local file has a different modified time 2025-May-09 09:28:13Z (UTC) when compared to remote modified time 2025-May-09 14:15:15Z (UTC)**_
```
```
2025-05-12 08:58:02,616 [OneDriveGUI.py:2022][fn=read_stdout][INFO] - [my user] The local file has a different hash when compared to remote file hash
2025-05-12 08:58:02,616 [OneDriveGUI.py:2022][fn=read_stdout][INFO] - [my user] Remote item modified time is newer based on UTC time conversion
2025-05-12 08:58:02,616 [OneDriveGUI.py:2022][fn=read_stdout][INFO] - [my user] The local item is out-of-sync with OneDrive, renaming to preserve existing file and prevent local data loss: file_name
2025-05-12 08:58:03,070 [OneDriveGUI.py:2022][fn=read_stdout][INFO] - [my user] Processing OneDrive JSON item batch [2/4] to ensure consistent local state
```
maybe some one have different time zone ?

**Reply by abraunegg** [reply]
No matter timezone, all timestamps are compared as UTC - so timezone is irrelevant.

Your logs illustrate a different hash .. meaning the files are different based on content - thus - this is why the safe backup is being created.

To ensure you are not being hit by any other issue, please build your client manually from 'master' and read the following:

* https://github.com/abraunegg/onedrive/blob/master/docs/usage.md#compatibility-with-editors-and-applications-using-atomic-save-operations

* https://github.com/abraunegg/onedrive/blob/master/docs/usage.md#compatibility-with-obsidian

**Comment by peterborto** [comment]
Hi @abraunegg  i've set in config :
```
force_session_upload = "true"
delay_inotify_processing = "true"
inotify_delay = "15"
```
but it tell me config is not valid.

**Reply by abraunegg** [reply]
@peterborto 
You will need to build your client from 'master' to use those options as they were introduced post v2.5.5 release.

v2.5.6 release will land shortly, however there are a number of important items to complete before I will commit to doing a new release.
